### DISCUSSION TITLE
Sixel Support

### DISCUSSION THREAD

**Discussion by rockorager** [ghostty-org/ghostty#2496] [Feature Requests, Ideas] [answered]
Sixels are the most commonly implemented image protocol to this day. It is [widely](https://www.arewesixelyet.com/) supported and often requested. Nevertheless, it is not without flaws. Let this be the canonical discussion on the merits of sixel support.

**Comment by rockorager** [comment]
I am firmly against supporting sixels at the terminal level. It is an antiquated protocol, designed around a technology that no longer exists and does not hold relevance to modern technology. Ghostty should actively support alternative image protocols. In my humble opinion, the Kitty graphics protocol is the current leader and all graphics support should push that direction.

In my opinion, we should actively discourage sixel support in terminals and instead encourage developers to use the kitty graphics protocol.

**Reply by tristan957** [reply]
Are there performance implications for the rest of ghostty if support were added?

**Reply by rockorager** [reply]
I'm not sure what the internal model would require for sixel support. It might not be *too* much overhead - it will have to detect an escape sequence, allocate the payload, decode the payload, and then decode that into pixels. Perhaps it could piggy-back on Kitty internally for the origin of an image, but for scrolls and redraw I imagine it will have some cost, because cursor placement comes into play with sixels. Perf costs are probably something that would need to be studied: it's not obvious how much of a cost it would be.

**Reply by norbusan** [reply]
Late to the game, but I didn't have access to the prerelease versions.

Performance if kitty protocol is rather bad compared to sixels. I get a 3x slowdown from 0.2 to 0.6sec on average.
I definitely will not use ghostty since I use image viewing in the console via mutt, and with ghostty this is painful.

**Reply by tristan957** [reply]
The Kitty Image Protocol performance is a known issue. Just needs someone to come along and work on it.

**Comment by jcollie** [comment]
I've played with implementing it a bit but never got as far as getting any images displayed mainly because I ran out of time and haven't gotten back to it.  It shouldn't take a whole lot more for someone to finish it off.  It certainly does have limitations, but _not_ implementing it because it's antiquated is kinda like saying that you shouldn't implement GIF support because animated PNGs are a thing.

Maybe if there was some kind of intermediary that would convert Sixel graphics to Kitty graphics we could get away with not supporting it...

**Comment by rockorager** [comment]
On Fri Oct 25, 2024 at 12:54 PM CDT, Jeffrey C. Ollie wrote:
> I've played with implementing it a bit but never got as far as getting any
> images displayed mainly because I ran out of time and haven't gotten back to
> it.  It shouldn't take a whole lot more for someone to finish it off.  It
> certainly does have limitations, but _not_ implementing it because it's
> antiquated is kinda like saying that you shouldn't implement GIF support
> because animated PNGs are a thing.
>
> Maybe if there was some kind of intermediary that would convert Sixel graphics
> to Kitty graphics we could get away with not supporting it...

I think it's antiquated legacy is the reason we shouldn't implement it. It can't
convey intent in the way we want to in today's environment. Sixels transmit an
image, and if part of it is clipped, the *client* has to re-encode the image to
the clipped region. This is a huge problem not only for ssh clients, but for
clients who just don't care about the viewable region. The kitty protocol was
designed to solve these issues, and to reach back to try to support a
sixel-pixel printer and force it to work on a modern terminal brings in too many
tradeoffs. We should appreciate this tech for what it is, and move on. We have
the technology, we have the protocol, there is no reason to provide a lesser
experience in 2024.

-- 
Tim

**Comment by tristan957** [comment]
> I'm not sure what the internal model would require for sixel support. 
> It might not be *too* much overhead - it will have to detect an escape 
> sequence, allocate the payload, decode the payload, and then decode 
> that into pixels. Perhaps it could piggy-back on Kitty internally for 
> the origin of an image, but for scrolls and redraw I imagine it will 
> have some cost, because cursor placement comes into play with sixels. 
> Perf costs are probably something that would need to be studied: it's 
> not obvious how much of a cost it would be.

Only reason I ask is because if there is no overhead and someone were 
willing to maintain it, I don't see a reason not to include support, but 
I'm not the guy maintaining this stuff, so my opinion is fairly moot.

-- 
Tristan Partin
https://tristan.partin.io

**Comment by rockorager** [comment]
On Fri Oct 25, 2024 at 3:27 PM CDT, Tristan Partin wrote:
> Only reason I ask is because if there is no overhead and someone were 
> willing to maintain it, I don't see a reason not to include support, but 
> I'm not the guy maintaining this stuff, so my opinion is fairly moot.

Yeah, I understand that sentiment. The issue I see is that sixel is
underspecified itself which makes implementation a moving target.
[Here](https://github.com/hpjansson/chafa/issues/192) is a lengthy discussion on
at least one of the issues.

-- 
Tim

**Comment by tristan957** [comment]
On Fri Oct 25, 2024 at 3:36 PM CDT, Tim Culverhouse wrote:
> On Fri Oct 25, 2024 at 3:27 PM CDT, Tristan Partin wrote:
>> Only reason I ask is because if there is no overhead and someone were 
>> willing to maintain it, I don't see a reason not to include support, but 
>> I'm not the guy maintaining this stuff, so my opinion is fairly moot.
>
> Yeah, I understand that sentiment. The issue I see is that sixel is
> underspecified itself which makes implementation a moving target.
> [Here](https://github.com/hpjansson/chafa/issues/192) is a lengthy discussion on
> at least one of the issues.

Great point. Then I think not supporting it makes sense if it will cause 
headaches for maintainers.

-- 
Tristan Partin
https://tristan.partin.io

**Comment by mitchellh** [comment, marked as answer]
My stance for now is that Ghostty will not support sixels.

This is a departure from our general pattern of supporting all manner of arcana from terminals past. It is also a departure from our pattern of supporting things xterm has with few questions. But those patterns are just that: patterns, they aren't law.

In this case, I'm making an informed decision based on my own research plus feedback here that sixels aren't worth it. They have numerous unspecified edge cases (some noted in this thread), the libsixel library is poorly written and poorly suited for drop-in adoption, the performance impacts are unclear but unlikely zero, and there is a feature rich future protocol that Ghostty already supports (the Kitty Graphics Protocol).

This decision isn't _final_ (as circumstances change, so do my opinions indefinitely), but I think it's definitive enough to answer this topic. I welcome any further discussion and feedback about it.

**Comment by Kreijstal** [comment]
someone fork ghostty with sixels?

**Comment by AndydeCleyre** [comment]
I'd like to add to this discussion the fact that tmux supports sixels but not the other image protocols, with AFAIK no plans to do so either. 

So for tmux users, Ghostty will effectively lack image support until/unless it gains sixel support. 

And while alternatives to tmux exist, for various reasons many continue to require it or prefer it to any of those existing alternatives.

**Reply by mitchellh** [reply]
> So for tmux users, Ghostty will effectively lack image support until/unless it gains sixel support.

This is not true. Kitty graphics supports the unicode placeholder feature exactly to satisfy tmux users. Kitty and Ghostty are the only two terminal emulators to support it: https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders

**Reply by AndydeCleyre** [reply]
Thanks! I'll read up on that and see if I can configure things to get it working, as it doesn't "out of the box" with the tools I'm using.

**Reply by joshbode** [reply]
I can confirm that Ghostty + tmux works perfectly with Kitty graphics :)

Try [timg](https://github.com/hzeller/timg)

<img width="489" alt="image" src="https://github.com/user-attachments/assets/341c9d95-8f22-4e38-8e30-f473c9ea835a" />

**Reply by l4zygreed** [reply]
how can I preview images in ranger using kitty protocol? 
if set preview method to kitty it's just show an empty preview 🤔

**Reply by rhodes-b** [reply]
> how can I preview images in ranger using kitty protocol? 
> if set preview method to kitty it's just show an empty preview 🤔 

This is not the place to ask make a new discussion in the future. Ranger recently merged support for all terminals with kip support but I assume thats not released into a version yet for now you have to open ranger with `TERM=xterm-kitty ranger` since it hardcoded only kitty supported it

**Reply by Kreijstal** [reply]
just use yazi

**Comment by sxyazi** [comment]
I think terminal multiplexers are still a valid argument for supporting Sixel.

The Kitty graphics protocol is feature-rich, well-defined, but also quite complex, this complexity affects its adoption speed. Currently, the two popular terminal multiplexers, tmux and Zellij, do not support it, but they both support Sixel:

- **tmux**: Since Ghostty supports Kitty unicode placeholders, it's possible to preview graphics in tmux via passthrough. But passthrough like a hack, it introduces other issues, such as not working in nested tmux sessions (both local and server run tmux and preview graphics on the server).
- **Zellij**: Zellij currently lacks passthrough, which prevents applications from previewing graphics in Ghostty.

**Reply by AnonymouX47** [reply]
You should check the actual state of sixel support in these multiplexers. Yes, they do advertise sixel support but how good is it in various scenarios? (I'm saying this not out of speculation but personal experience)

Then, take that and compare it to what you get via kitty graphics' unicode placeholders.

Also, consider this: In a remote tmux session, scroll the terminal with an image on screen and compare the amount of bytes transferred over the network and the redraw performance.

To be honest, instead of pushing for sixel support here, anyone should be pushing for passthrough support in Zellij (and any other multiplexer that lacks it).

- - -

For what's worth, nothing stops passthrough from working in arbitrarily nested sessions. You only need to nest the passthrough sequence accordingly. As for detecting nested sessions, `XTVERSION` queries wrapped in passthrough sequences should pull it off.

**Reply by sxyazi** [reply]
My intention is not to push anything; I was just passing by to add some supplements to the previous discussion. If these contributions are unhelpful and create noise, I apologize.

I don't think it's very meaningful to compare Sixel and Kgp here. This discussion is about “adding extra Sixel support”, not “using Sixel to replace the existing Kgp”.

Adding Sixel means that both users and developers have a new option available. If one must compare, in my view: Kgp > Sixel > Kgp via passthrough.

> You only need to nest the passthrough sequence accordingly. As for detecting nested sessions, `XTVERSION` queries wrapped in passthrough sequences should pull it off.

This is the crux of the problem. The existing Unicode placeholders need to know the tmux nesting level and nest the passthrough sequence accordingly, which requires querying `XTVERSION` via passthrough — but that is not what passthrough was designed for.

According to the tmux maintainer, https://github.com/tmux/tmux/issues/4386#issuecomment-2681737611:

> You can't realistically use passthrough for anything that requires a response, that is not what it is intended for.

And once an application starts using passthrough to receive responses, it's like opening Pandora's box: you have to handle a series of issues it brings along:

- tmux cannot guarantee the order of passthrough responses. If a request contains both passthrough and non-passthrough sequences simultaneously, the order of responses is unpredictable. You could split them into two separate requests, and wait for the response of each previous request in order, but that would significantly increase the startup time in “non-standard” environments (see point 3).
- Passthrough does not work in synced tmux panes (for example, one terminal running `tmux` and another using `tmux attach`), which results in a single request being responded to twice by two different terminals and disrupting the app.
- Passthrough is not always available, and there is no reliable way to detect its availability. For instance, it does not work under `tmux -CC` or `tmux popup`. Application developers must be prepared at all times for the possibility of receiving no response. This necessitates many timeout fallback mechanisms, and the timeout cannot be set too short to account for the network uncertainties of (normal) SSH sessions, leading to significantly slower startup times in these “non-standard” environments.
- If in the future Zellij supports passthrough, these issues might apply as well — or even worsen — because applications would then need to support another terminal multiplexer in addition to tmux.

**Reply by AnonymouX47** [reply]
> My intention is not to push anything; I was just passing by to add some supplements to the previous discussion. If these contributions are unhelpful and create noise, I apologize.

Oh, not at all. I apologise for my tone earlier. :pray:

- - -

To be clear, my major concern is whether the maintenance burden is worth what we get out of it.

Thank you so much for the corrections and clarifications about using passthrough.

**Reply by sxyazi** [reply]
No worries! I understand the concerns about maintenance burden — I run some open source projects myself, and sometimes I have to make trade-offs and drop certain features to avoid making maintenance an endless headache. 

That said, if there ever comes a day when Sixel support is reconsidered and my input helps in the decision-making, I'd be truly honored. 😉

**Comment by pmarreck** [comment]
I am disappointed that Ghostty doesn't support Sixel.

That's all I feel I need to say on the matter.
